src/common/vitals/vital-metric.js

```diff
@@ -0,0 +1,55 @@
+export class VitalMetric {
+  #values = []
+  #subscribers = new Set()
+  entries = []
+
+  constructor (name, roundingMethod) {
+    this.name = name
+    this.attrs = {}
+    this.roundingMethod = typeof roundingMethod === 'function' ? roundingMethod : Math.floor
+  }
+
+  update ({ value, entries = [], attrs = {}, addConnectionAttributes = false }) {
+    this.#values.push(this.roundingMethod(value))
+    this.entries.push(...entries)
+    this.attrs = { ...this.attrs, ...attrs }
+    if (addConnectionAttributes) this.addConnectionAttributes()
+    this.#subscribers.forEach(cb => {
+      try {
+        cb(this.value)
+      } catch (e) {
+        // ignore errors
+      }
+    })
+  }
+
+  get value () {
+    return {
+      previous: this.#values[this.#values.length - 2],
+      current: this.#values[this.#values.length - 1],
+      name: this.name,
+      attrs: this.attrs,
+      entries: this.entries
+    }
+  }
+
+  get isValid () {
+    return this.#values.length > 0 && this.value.current >= 0
+  }
+
+  subscribe (callback) {
+    this.#subscribers.add(callback)
+    if (this.isValid) callback(this.value)
```

Comment:
If there's already data (prev calls to `.update`), you'd want to cb with _each_ of those value-entries.

Reply:
not following what you mean here, `this.value` is a getter which returns previous, current, and entries (all the values seen). Is that what you mean?
```javascript
get value () {
    return {
      previous: this.#values[this.#values.length - 2],
      current: this.#values[this.#values.length - 1],
      name: this.name,
      attrs: this.attrs,
      entries: this.entries
    }
  }
```

https://github.com/newrelic/newrelic-browser-agent/pull/635/files/7b056dd4bb8c7c1b9a78ecb9bc015c78d57fa2e1#diff-51f5c7465d0ec05605d24c9d75d2580ee046be3b14756d95033e85c7ceeed67cR26-R34

Reply:
Well suppose `this.#values = [1, 5, 2, 4]` when `subscribe` is called. It runs for the current value but none of the pre-existing. This differs from what observers with buffered would do. In my eyes, `subscribe` being for only the most current "updates" and future updates but not past updates is not a common pattern; parallel to the EE not executing the `.on(___,cb)` with the most recent emitted same-type event & args. 

I think to sum it up, this could be confirmed to be either _only_ forward-looking (like ee.on) or both forward and **all** backwards looking (like PO.observer buffered)

Reply:
this was somewhat patterned off of rxjs's behaviorSubject, which will emit the latest value when subscribed to https://stackblitz.com/edit/rxjs-behaviorsubject-simpleexample?devtoolsheight=100&file=index.ts . Ours just has extra metadata attached as well on that emission.

if that's confusing, we can always iterate on this pattern tho

Reply:
Not really though, because it also reports the entries, which contains the full history of the emissions

Reply:
Just to be clear -- Basically it always reports the full history (entries), but the returned object also has shortcut methods the get the latest value or previous value

Reply:
If a cwv emits (valueA derived from setA entries) and (valueB derived from setB entries) and (valueC from setC), it doesn't make sense you can just report valueC and [...setA, ...setB, ...setC] (can't even differentiate the line between the sets) and expect the receiver to reconstruct valueA & valueB.

Reply:
changed to include all updates as history objects, see: https://github.com/newrelic/newrelic-browser-agent/pull/635/commits/39fc1e91a53329c4c2609d030e4f31e0f64cc762